### ISSUE TITLE
Starting PropertyOf producer multiple times

### ISSUE THREAD

**Issue by harrywincup** [ReactiveCocoa/ReactiveCocoa#2324]
Could somebody please explain my misunderstanding of the following:

```swift
let subscriptionProducer = subscriptionService.fetch()
	|> on(started: { _ in println("subscriptionProducer STARTED") })
	|> catch { _ in SignalProducer.empty }
	|> flatMap(.Latest, subscriptionViewModelProducer)

subscriptions = PropertyOf(initialValue: [], producer: subscriptionProducer)

subscriptions.producer.start(next: { _ in
	println("subscriptions.producer.start 1")
})

subscriptions.producer.start(next: { _ in
	println("subscriptions.producer.start 2")
})

subscriptions.producer.start(next: { _ in
	println("subscriptions.producer.start 3")
})
```

I would expect `subscriptionProducer STARTED` to be printed 4 times, once for the implicit `start` when creating the `PropertyOf` and once for each subsequent `subscriptions.producer.start`, but I see it only once.

I assumed that I should call `subscriptions.producer.start` whenever I want to refresh the data but now i'm not sure if that's a stupid idea.

**Comment by neilpa** [comment]
[Properties](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/554364022b83bb6d67eaf8354956a52c02236b11/Documentation/FrameworkOverview.md#properties) start the internal producer once and cache the most recent value. If you actually want to start the producer multiple times why are you using a property?

**Comment by harrywincup** [comment]
@neilpa I would say the answer to that is because I have no idea what i'm doing :) 

I'm trying to use the `subscriptions` property on my view model to populate the collection view of the related controller:

```swift
viewModel.producer
	|> ignoreNil
	|> flatMap(.Latest) { $0.subscriptions.producer }
	|> start(next: { [unowned self] _ in
		self.collectionView?.reloadData()
	})
``` 

Then I was attempting to update the collection view when new subscriptions are added ... but I'm not sure how to get those, so I tried to start the producer again, which led me down this rabbit hole.

Could you give me a pointer on what I should be using instead of `PropertyOf` please? Or if my whole approach is wrong, please point me in the right direction.

Thank you!

**Comment by neilpa** [comment]
Sadly there's no documentation yet about holistic app building the "RAC3 way". I think that's the underlying question that you're asking. I've done some experiments and hope to some point document them but I'm not there yet.

> Then I was attempting to update the collection view when new subscriptions are added

Depends on who controls this and when/how you want the data refreshed. For your case you probably want whoever controls the subscription refresh to have an internal `MutableProperty<[Subscription]>` that backs a public `PropertyOf<[Subscription]>`.

**Comment by harrywincup** [comment]
Yeah, some real life examples of using RAC3 would be super useful. I've read all of the existing documents a lot of times, but i've found they're difficult to piece together when you don't know what you don't know.

> For your case you probably want whoever controls the subscription refresh to have an internal MutableProperty<[Subscription]> that backs a public PropertyOf<[Subscription]>.

Interesting, I thought the purpose of the `PropertyOf(initialValue:producer:)` initializer was that you no longer needed to keep a separate `MutableProperty` as state (because one is set up internally by the property)

It looks like I still have a long way to go before this stuff sinks in. Thanks for the pointer!

**Comment by neilpa** [comment]
> Interesting, I thought the purpose of the PropertyOf(initialValue:producer:) initializer was that you no longer needed to keep a separate MutableProperty as state (because one is set up internally by the property)

Yes, but it still only `start`s the provided producer once, caching the most recent value. In your case your producer looks like a network request that sends the results once then completes. If you instead had some other signal/producer that always sent the most recent set of `[Subscription]` you could use `PropertyOf` directly.

**Comment by harrywincup** [comment]
Yep, it's definitely the caching that has confused me (incidentally, I thought that's what `ConstantProperty` did, which is perhaps yet another misunderstanding!).

I'll try hooking up the refresh to a different producer and feeding that to the property. Thanks for taking time to explain what are probably very obvious things. As a newbie, I really appreciate it :)

**Comment by neilpa** [comment]
This is why I want to put together an end-2-end guide eventually. It's a different style of thinking when writing apps. And it's not always clear the best way to use RAC to achieve it.

**Comment by harrywincup** [comment]
I absolutely look forward to reading it. Seeing how the experts approach something will drastically reduce the learning curve for this stuff!

**Comment by neilpa** [comment]
You may also be interested in [Rex](https://github.com/neilpa/rex). It's where I've done some experimenting with improving RAC3 and Cocoa interop. There's not a lot there yet though since I add stuff as I need it.

**Comment by harrywincup** [comment]
@neilpa Apologies for coming back to this, but after having more of a play i'm still very confused about how the properties work. I think my use case is very typical ... 

* I have a collection view controller, backed by a view model. 
* The view model has a producer, which makes a core data request and sends the results via `sendNext`, then immediately completing via `sendCompleted`
* The view model has a public property (previously `PropertyOf`, now trying `MutableProperty`), which is used as the data source for the collection view: `return subscriptions.value.count`
* I feed the producer values into the property with `subscriptions <~ subscriptionProducer`
* The view controller needs to know when the property changes so it can `reloadData()`

If I start the `subscriptionProducer`, should `subscriptions` be updated with the new values once the producer completes? ... Even with a `MutableProperty`, the value seems to remain the same.

How should I be observing the changes to `subscriptions` in my view controller so I can reload the collection view?

This seems like it should be a very simple thing to do, but I can't find any examples

**Comment by harrywincup** [comment]
Ok, I've tried a different approach with this and am now using an `Action` on my view model to start the producer that fetches data. It seems that Action exposes a signal, which I can observe from my view controller in order to reload the collection view when new values come in.

I have no idea if this is closer to the recommended approach, but at least it seems to get me nearer to what i'm trying to do

**Comment by neilpa** [comment]
Using `Action` sounds reasonable since you probably want serialized refresh behavior with shared results. Sorry I couldn't point you in the "right" direction faster.
